kernel/arch/dreamcast/include/dc/g2bus.h

```diff
@@ -31,50 +33,165 @@
+    These functions will be called inside an interrupt context, so don't try to
+    use anything that might stall.
+
+    \param  data            User data passed in to the g2_dma_transfer()
+                            function.
+*/
+typedef void (*g2_dma_callback_t)(void *data);
+
+/** \brief  Perform a DMA transfer between SH-4 RAM and G2 Bus
+
+    This function copies a block of data between SH-4 RAM and G2 Bus via DMA. 
+    You specify the direction of the copy (SH4TOG2BUS/G2BUSTOSH4). There are all 
+    kinds of constraints that must be fulfilled to actually do this, so
+    make sure to read all the fine print with the parameter list.
+
+    If a callback is specified, it will be called in an interrupt context, so
+    keep that in mind in writing the callback.
+
+    \param  sh4             Where to copy from/to. Must be 32-byte aligned.
+    \param  g2bus           Where to copy from/to. Must be 32-byte aligned.
+    \param  length          The number of bytes to copy. Must be a multiple of
+                            32.
+    \param  block           Non-zero if you want the function to block until the
+                            DMA completes.
+    \param  callback        A function to call upon completion of the DMA.
+    \param  cbdata          Data to pass to the callback function.
+    \param  dir             SH4TOG2BUS or G2BUSTOSH4.
+    \param  mode            Any value; Here for compatibility.
+    \param  g2chn           See g2b_channels.
+    \param  sh4chn          Any value; Here for compatibility.
+    \retval 0               On success.
+    \retval -1              On failure. Sets errno as appropriate.
+
+    \par    Error Conditions:
+    \em     EINPROGRESS - DMA already in progress \n
+    \em     EFAULT - sh4 and/or g2bus is not 32-byte aligned \n
+    \em     EINVAL - Invalid g2chn
+    \em     EIO - I/O error
+
+*/
+int g2_dma_transfer(void *sh4, void *g2bus, size_t length, uint32_t block,
+                    g2_dma_callback_t callback, void *cbdata,
+                    uint32_t dir, uint32_t mode, uint32_t g2chn, uint32_t sh4chn);
+
+/** \brief  Initialize DMA support.
+
+    This function sets up the DMA support for transfers to/from the G2 Bus.
 
-   Update : looks like there is a formula, mode = 3 + shchn
+    \retval 0               On success (no error conditions defined).
+*/
+int g2_dma_init(void);
+
+/** \brief  Shutdown DMA support. */
+void g2_dma_shutdown(void);
+
+/** \brief  G2 context
+
+    A G2 context containing the states of IRQs and G2 DMA. This struct
+    is used in with g2_lock() and g2_unlock().
+*/
+typedef struct { 
+    uint32_t irq_state;    /** \brief IRQ state when entering a G2 critical block */
+} g2_ctx_t;
+
+/* Internal constants */
+/** \cond */ 
+#define SPU_G2DMA_SUSPEND     (*((vuint32 *)0xa05f781C))
+#define BBA_G2DMA_SUSPEND     (*((vuint32 *)0xa05f783C))
+#define CH2_G2DMA_SUSPEND     (*((vuint32 *)0xa05f785C))
```

Comment:
Can we rename these as well? `G2_DMA_SUSPEND_SPU`, `G2_DMA_SUSPEND_BBA`, `G2_DMA_SUSPEND_CH2` seem better to me. Also... It would be nice if these were documented if they're going to be in a public header.

Reply:
Renamed to your specifications